IdentityCore/src/cache/mac/MSIDMacSharedCredentialCacheItem.m

```diff
@@ -0,0 +1,190 @@
+        }
+    }
+    
+    return self;
+}
+
+- (NSDictionary *)jsonDictionary
+{
+    NSDictionary *cacheObjects = [self getCopy];
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    
+    for (MSIDDefaultCredentialCacheKey *refreshTokenKey in cacheObjects)
+    {
+        NSString *key = [NSString stringWithFormat:@"%@%@%@", refreshTokenKey.account, keyDelimiter, refreshTokenKey.service];
+        MSIDCredentialCacheItem *refreshToken = [cacheObjects objectForKey:refreshTokenKey];
+        NSDictionary *rtDict = [refreshToken jsonDictionary];
+        
+        if (rtDict)
+        {
+            [dictionary setObject:rtDict forKey:key];
+        }
+    }
+    
+    return dictionary;
+}
+
+- (void)setSharedCredential:(MSIDCredentialCacheItem *)token forKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:token forKey:key];
+    });
+}
+
+-(MSIDCredentialCacheItem *)sharedCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    __block MSIDCredentialCacheItem *sharedCredential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        sharedCredential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return sharedCredential;
+}
+
+- (void)removeSharedCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects removeObjectForKey:key];
+    });
+}
+
+
+- (void)mergeSharedCredential:(MSIDMacSharedCredentialCacheItem *)credential
+{
+    NSDictionary *copy = [credential getCopy];
+    
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects addEntriesFromDictionary:copy];
+    });
+}
+
+- (NSArray<MSIDCredentialCacheItem *> *)sharedCredentialsWithKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    if (key.account && key.service)
+    {
+        MSIDCredentialCacheItem *credential = [self sharedCredentialForKey:key];
+        if (credential)
+        {
+            return @[credential];
+        }
+        
+        return nil;
+    }
+    
+    NSMutableArray *subPredicates = [[NSMutableArray alloc] init];
+    
+    if (key.clientId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.clientId == %@", key.clientId]];
+    if (key.environment)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.environment == %@", key.environment]];
+    if (key.homeAccountId)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.homeAccountId == %@", key.homeAccountId]];
+    if (key.credentialType)
+        [subPredicates addObject:[NSPredicate predicateWithFormat:@"self.credentialType == %d", key.credentialType]];
+    
+    // Combine all sub-predicates with AND:
+    NSPredicate *matchAttributes = [NSCompoundPredicate andPredicateWithSubpredicates:subPredicates];
+    
+    NSDictionary *cacheObjects = [self getCopy];
+    return [[cacheObjects allValues] filteredArrayUsingPredicate:matchAttributes];
```

Comment:
A dispatch_sync() here around the call to filteredArrayUsingPredicate should eliminate the need for getCopy.

Reply:
Same comment as above. Dispatch_sync is already used in the copy method. Lets discuss this offline.